lib/configurators/databases/aws.go

```diff
@@ -0,0 +1,639 @@
+func buildIAMAutoDiscoveryStatements(target awslib.Identity) ([]*awslib.Statement, error) {
+	statement := &awslib.Statement{
+		Effect:    awslib.EffectAllow,
+		Resources: []string{target.String()},
+	}
+
+	switch target.(type) {
+	case awslib.User, *awslib.User:
+		statement.Actions = userBaseActions
+	case awslib.Role, *awslib.Role:
+		statement.Actions = roleBaseActions
+	default:
+		return nil, trace.BadParameter("policies target must be an user or role, received %T", target)
+	}
+
+	return []*awslib.Statement{statement}, nil
+}
+
+// buildRDSAutoDiscoveryStatements returns IAM statements necessary for
+// RDS/Aurora databases auto-discovery.
+func buildRDSAutoDiscoveryStatements() ([]*awslib.Statement, error) {
+	return []*awslib.Statement{
+		{
+			Effect:    awslib.EffectAllow,
+			Actions:   append(rdsActions, auroraActions...),
+			Resources: []string{"*"},
+		},
+	}, nil
+}
+
+// buildRDSAutoDiscoveryBoundaryStatements returns IAM boundary statements
+// necessary for RDS/Aurora databases auto-discovery.
+func buildRDSAutoDiscoveryBoundaryStatements() ([]*awslib.Statement, error) {
+	return []*awslib.Statement{
+		{
+			Effect:    awslib.EffectAllow,
+			Actions:   append(rdsActions, append(auroraActions, boundaryRDSAuroraActions...)...),
+			Resources: []string{"*"},
+		},
+	}, nil
+}
+
+// buildRedshiftAutoDiscoveryStatements returns IAM statements necessary for
+// Redshift databases auto-discovery.
+func buildRedshiftAutoDiscoveryStatements() ([]*awslib.Statement, error) {
+	return []*awslib.Statement{
+		{
+			Effect:    awslib.EffectAllow,
+			Actions:   redshiftActions,
+			Resources: []string{"*"},
+		},
+	}, nil
+}
+
+// buildRedshiftAutoDiscoveryBoundaryStatements returns IAM boundary statements
+// necessary for Redshift databases auto-discovery.
+func buildRedshiftAutoDiscoveryBoundaryStatements() ([]*awslib.Statement, error) {
+	return []*awslib.Statement{
+		{
+			Effect:    awslib.EffectAllow,
+			Actions:   append(redshiftActions, boundaryRedshiftActions...),
+			Resources: []string{"*"},
+		},
+	}, nil
+}
+
+// buildRedshiftStatements generates policy Statements for Redshift databases.
+func buildRedshiftStatements(fileConfig *config.FileConfig, accountID string) ([]*awslib.Statement, error) {
+	var resources []string
+	for _, database := range fileConfig.Databases.Databases {
+		if strings.Contains(database.URI, types.RedshiftEndpointSuffix) {
+			clusterID, region, err := types.ParseRedshiftEndpoint(database.URI)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			for _, resource := range redshiftResources {
+				resources = append(resources, fmt.Sprintf(
+					"arn:aws:redshift:%s:%s:%s:%s/*",
+					region,
+					accountID,
+					resource,
+					clusterID,
+				))
+			}
+		}
+	}
+
+	if len(resources) == 0 {
+		return []*awslib.Statement{}, nil
+	}
+
+	return []*awslib.Statement{
+		{Effect: awslib.EffectAllow, Actions: []string{"redshift:GetClusterCredentials"}, Resources: resources},
```

Comment:
Hmm, do we need to include this in the policy? This is what Teleport agent creates itself during auto-IAM configuration, no? And the `resources` loop above probably isn't needed also.

Reply:
You're right. This is similar to the one Teleport generates. We still need to place them on the boundary policy, or Teleport won't have enough permissions to attach to its user/role. And this loop is reducing the number of "wildcard permissions" we could place a default wildcard permission (like the one added when Redshift auto-discovery is enabled).

Reply:
> Hmm does that mean though that they will have to re-run configurator after adding each new database?
For static Redshift database, yes. It is the tradeoff of having a more restrictive boundary.

We can identify that the agent has dynamic registration and go for the wildcard in those cases since we don't know the complete databases list. And since it is the recommended way, going with only wildcard permissions for Redshift sounds ok.

Reply:
Yes, let's do that. Do you need to update this method then to remove the loop above?